text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,581 @@
+  demonstration(list: string[] | number[]): void;
+}
+```
+
+Cannot be expressed in a satisfactory way in **Java**:
+
+```java
+public interface IErasureProblem {
+  public void demonstration(final List<String> list);
+
+  // error: name clash: demonstration(List) and demonstration(List) have the same erasure
+  public void demonstration(final List<Integer> list);
+}
+```
+
+Instead, it must be rendered as:
+
+```java
+public interface IErasureProblem {
+  public void demonstration(final List<?> list);
+}
+```
+
+[reification]: https://en.wikipedia.org/wiki/Reification_(computer_science)
+[type erasure]: https://docs.oracle.com/javase/tutorial/java/generics/erasure.html
+
+### Impossible to express collections of a _type union_
+
+Similar to the issue with generating method overrides, there is no way in statically typed languages to express a collection of a _type union_.
+
+```java
+public class CfnAlias {
+  // ...
+  public interface AliasRoutingConfigurationProperty {
+    // ...
+    public interface Builder extends Builder<AliasRoutingConfigurationProperty> {
+      // ...
+      // Impossible to express Array<CfnAlias.VersionWeightProperty | cdk.IResolvable> in java
+      public Builder additionalVersionWeights(final List<?> value);
+      public Builder additionalVersionWeights(final IResolvable value);
+      // ...
+    }
+    // ...
+  }
+  // ...
+}
+```
+
+### Bugs in the Kernel
+
+Currently, the _jsii Kernel_ protocol requires each object instance is annotated with exactly one type, and this type is expected to match the dynamic
+type of the instance. Concretely, this means the value returned from **Javascript** to **Java** via a return type annotated as a _type union_ is
+expected to be one of the union candidates.
+
+While this poses no problems in the case of unions of classes known to _jsii_ (those classes that are exported from a _jsii library_, and hence have
+an assigned _jsii FQN_) can be annotated with their _real_ dynamic type without problems. However in the case where the returned value is an object
+literal (either a struct literal, or an anonymous interface implementation), the **Javascript** runtime provides no way to determine which struct or
+interface of the _type union_ candidates is implemented by the value.
+
+Taking the example of `AWS::SAM::Function`'s `events` property, the union of possible event sources types only includes structs, and one of the
+candidates, `AlexaSkill` has no required properties, meaning _any_ object actually satisfies it's specification.
+
+In those cases, the current _jsii kernel_ will actually annotate _all_ instances with the first candidate type it investigates (regardless of whether
+the instance conforms to it's specification or not). While schema validation could be performed to only use a particular option if the instance
+conforms to it's schema, it would not allow solving the ambiguity around items without any required property (rejecting the presence of additional
```

Comment:
It would somewhat. We could (probably, need to think on this) the types so that they are ordered from most to least restrictive.

In that case, the first one that matches is [probably] the intended type, with the most permissive "anything goes" type coming last and not accidentally swallowing the real type.

Reply:
The issue is that in case two types only have optional properties, a value that has no properties matches both. If an API has a contract such as "if foo is bar, then baz is <some type>", but the kernel elects the other empty-compatible candidate, then user will have a (bad) surprise.

It's not always possible to generate a dynamic proxy that represents "both" or "either" of those types (and the Kernel API definitely does not support such a wire type currently).

Reply:
We can disallow unions where it is impossible to discriminate between the types.

Anyway, I'd like to see more detail on this specific problem in the RFC as a whole. It's a little undertreated right now.

Reply:
I'd still like to see whether you consider this problem in scope or out of scope for the current RFC, and if in scope what the proposed handling of:

```
type PropsUnion = FooProps | BarProps;
```

Is going to be.

Reply:
This is documented... You'll get the union in the same way than you would in **TypeScript**, and you can either use a type inspection API (`isFooProps(v: PropsUnion): v is FooProps`) and you can "safely" force it, or you don't have this and you have to opportunistically force it... Either way you get similar ergonomics & safety to what you have in **TypeScript** in the same scenario.